alacritty_terminal/src/keyboard_motion.rs

```diff
@@ -0,0 +1,567 @@
+                point
+            } else if left {
+                term.semantic_search_left(point)
+            } else {
+                term.semantic_search_right(point)
+            }
+        };
+
+        let mut buffer_point = term.visible_to_buffer(point);
+
+        // Move to word boundary
+        if !is_boundary(term, buffer_point, left) && left != start {
+            buffer_point = semantic(buffer_point);
+        }
+
+        // Skip whitespace
+        let mut point = advance(term, buffer_point, left);
+        while !is_boundary(term, buffer_point, left) && is_space(term, point) {
+            buffer_point = point;
+            point = advance(term, buffer_point, left);
+        }
+
+        // Assure minimum movement of one cell
+        if !is_boundary(term, buffer_point, left) {
+            buffer_point = advance(term, buffer_point, left);
+        }
+
+        // Move to word boundary
+        if !is_boundary(term, buffer_point, left) && left == start {
+            buffer_point = semantic(buffer_point);
+        }
+
+        // Scroll viewport if necessary
+        Self::scroll(term, buffer_point);
+
+        term.buffer_to_visible(buffer_point).unwrap_or_else(Point::default).into()
+    }
+
+    /// Move by whitespace separated word, like W/B/E/gE in vi.
+    fn word_move<T: EventListener>(
+        term: &mut Term<T>,
+        point: Point,
+        left: bool,
+        start: bool,
+    ) -> Point {
+        let mut buffer_point = term.visible_to_buffer(point);
+
+        // Skip whitespace until right before a word
+        if left == start {
+            let mut point = advance(term, buffer_point, left);
+            while !is_boundary(term, buffer_point, left) && is_space(term, point) {
+                buffer_point = point;
+                point = advance(term, point, left);
+            }
+        }
+
+        if left == start {
+            // Skip non-whitespace until right inside word boundary
+            let mut point = advance(term, buffer_point, left);
+            while !is_boundary(term, buffer_point, left) && !is_space(term, point) {
+                buffer_point = point;
+                point = advance(term, buffer_point, left);
+            }
+        } else {
+            // Skip non-whitespace until just beyond word
+            while !is_boundary(term, buffer_point, left) && !is_space(term, buffer_point) {
+                buffer_point = advance(term, buffer_point, left);
+            }
+        };
+
+        // Skip whitespace until right inside word boundary
+        if left != start {
+            while !is_boundary(term, buffer_point, left) && is_space(term, buffer_point) {
+                buffer_point = advance(term, buffer_point, left);
+            }
+        }
+
+        // Scroll viewport if necessary
+        Self::scroll(term, buffer_point);
+
+        term.buffer_to_visible(buffer_point).unwrap_or_else(Point::default).into()
+    }
```

Comment:
I think the reason this is feeling a bit unwieldy is partly because we're overloading some functionality now. Before we had just normal and "semantic" selection. Now while those concepts remain, we add a few semantic-like features for just the keyboard. I think I'd rather see these be "kinds" of semantic selection, so it may be possible to use them with the mouse as well. This should hopefully lead to better code, and more consistent user experience.

Reply:
> Before we had just normal and "semantic" selection. Now while those concepts remain, we add a few semantic-like features for just the keyboard.

Well word motion has nothing to do with semantic selection at all really. And the punctuation is basically identical to semantic selection. So I don't see how you would use any of this for the mouse pointer.

Moving to the end of the previous word is pointless for selection expansion, but is very much necessary for the semantic selection, so I don't see how that would be helpful to have in the selection code?

Fundamentally the user experience is identical, the expansion of the keyboard motion punctuation movements and the selection expansion are 1:1 the same.

Reply:
Should punctuation be called semantic, or visa versa then? Can we share any code?

Reply:
We already share all the code we can share by calling semantic_search from the keyboard motion afaik. All the logic I added is specific to keyboard motion.